src/rubrix/server/tasks/token_classification/metrics.py

```diff
@@ -160,7 +161,100 @@ def aggregation_result(self, aggregation_result: Dict[str, Any]) -> Dict[str, An
         return {"mentions": result}
 
 
-class TokenClassificationMetrics(BaseTaskMetrics):
+class F1Metric(PythonMetric[TokenClassificationRecord]):
+    """The F1 metric based on entity-level.
+
+    We follow the convention of `CoNLL 2003 <https://aclanthology.org/W03-0419/>`_, where:
+    `"precision is the percentage of named entities found by the learning system that are correct.
+    Recall is the percentage of named entities present in the corpus that are found by the system.
+    A named entity is correct only if it is an exact match (...).”`
+    """
+
+    def apply(self, records: Iterable[TokenClassificationRecord]) -> Dict[str, Any]:
+        # store entities per label in dicts
+        predicted_entities = {}
+        annotated_entities = {}
+
+        # extract entities per label to dicts
+        for rec in records:
+            if rec.prediction:
+                self._add_entities_to_dict(rec.prediction.entities, predicted_entities)
+            if rec.annotation:
+                self._add_entities_to_dict(rec.annotation.entities, annotated_entities)
+
+        # store precision, recall, and f1 per label
+        per_label_metrics = {}
+
+        annotated_total, predicted_total, correct_total = 0, 0, 0
+        precision_macro, recall_macro = 0, 0
+        for label, annotated in annotated_entities.items():
+            predicted = predicted_entities.get(label, set())
+            correct = len(annotated & predicted)
+
+            # safe divides are used to cover the 0/0 cases
+            precision = self._safe_divide(correct, len(predicted))
+            recall = self._safe_divide(correct, len(annotated))
+            per_label_metrics.update(
+                {
+                    f"{label}_precision": precision,
+                    f"{label}_recall": recall,
+                    f"{label}_f1": self._safe_divide(
+                        2 * precision * recall, precision + recall
+                    ),
+                }
+            )
+
```

Comment:
Not sure if have sense group metrics by kind (micro/macro and per_label) instead set all as flat dictionary:

```python
{
   "micro": { "recall",... "precision"....},
   "macro": { "recall",... "precision"....},
   "per_label": {
      "a": { "recall",... "precision"....},
      "b": { "recall",... "precision"....}
   }
```

what do you think, @dcfidalgo ?

Reply:
This was actually my first schema, but then I thought it would be more practical to flatten it for further usage:
- quicker access: `summary.data["MyLabel_f1"]` versus `summary.data["per_label"]["MyLabel"]["f1"]`
- summary table: `pd.DataFrame(summary.data.values(), index=summary.data.keys())` versus `pd.DataFrame(.........)`